google/cloud/storage/status_or.h

```diff
@@ -0,0 +1,178 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_STATUS_OR_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_STATUS_OR_H_
+
+#include "google/cloud/internal/throw_delegate.h"
+#include "google/cloud/storage/internal/throw_status_delegate.h"
+#include "google/cloud/storage/status.h"
+#include <type_traits>
+#include <utility>
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+/**
+ * `StatusOr<T>` is used to return a value or an error status when exceptions
```

Comment:
I think we may want to remove most of the "when exceptions are disabled" comments on this class. `StatusOr<T>` is used to return a T or an error/status, regardless of whether exceptions are enabled or not. The only thing this class cares about exceptions is taht it's .value() will likely abort() if exceptions aren't enabled; otherwise, .value() should throw.

Reply:
Done.